src/extension/serializer.ts

```diff
@@ -1009,16 +1047,395 @@ export class GrpcSerializer extends SerializerBase {
+    //   (c) other, manually configured
+
+    // I will assume 2a here
+    const tlsPath = path.normalize(__dirname + '/../tls')
+    const keyPath = `${tlsPath}/key.pem`
+    const certPath = `${tlsPath}/cert.pem`
+
+    let grpcOptions = {
+      baseUrl: this.serializerServiceUrl,
+      httpVersion: '2',
+      nodeOptions: {
+        rejectUnauthorized: false, // allow self-signed certificates
+        cert: fs.readFileSync(certPath),
+        key: fs.readFileSync(keyPath),
+      },
+    } as GrpcTransportOptions
+
+    return createPromiseClient(ParserService, createGrpcTransport(grpcOptions))
+
+    // return createConnectTransport({
+    //   baseUrl: this.serializerServiceUrl,
+    //   httpVersion: '1.1',
+    // })
   }
 
-  public getNotebookDataCache(cacheId: string): NotebookData | undefined {
-    return this.notebookDataCache.get(cacheId)
+  protected async saveNotebook(
+    data: NotebookData,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    token: CancellationToken,
+  ): Promise<Uint8Array> {
+    const marshalFrontmatter = this.lifecycleIdentity === RunmeIdentity.ALL
+
+    const notebook = ConnectSerializer.marshalNotebook(data, { marshalFrontmatter })
+
+    if (marshalFrontmatter) {
+      data.metadata ??= {}
+      data.metadata[RUNME_FRONTMATTER_PARSED] = notebook.frontmatter
+    }
+
+    const cacheId = GrpcSerializerBase.getDocumentCacheId(data.metadata)
+    this.notebookDataCache.set(cacheId as string, data)
+
+    const serialRequest = new es_proto.SerializeRequest({ notebook })
+
+    const cacheOutputs = this.cacheNotebookOutputs(notebook, cacheId)
+    const request = this.client.serialize(serialRequest)
+
+    // run in parallel
+    const [serialResult] = await Promise.all([request, cacheOutputs])
+
+    if (cacheId) {
+      await this.saveNotebookOutputsByCacheId(cacheId)
+    }
+
+    if (serialResult.result === undefined) {
+      throw new Error('serialization of notebook failed')
+    }
+
+    return serialResult.result
+  }
+
+  protected async reviveNotebook(
+    content: Uint8Array,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    token: CancellationToken,
+  ): Promise<Serializer.Notebook> {
+    const identity = this.lifecycleIdentity
+    // const markdown = Buffer.from(content).toString('utf8')
+    const deserializeRequest = new es_proto.DeserializeRequest({
+      source: content,
+      options: { identity },
+    })
+    let res
+    try {
+      res = await this.client.deserialize(deserializeRequest)
+    } catch (e: any) {
+      if (e.name === 'ConnectError') {
+        e.message = `Unable to connect to the serializer service at ${this.serializerServiceUrl}`
+      }
+      log.error('Error in reviveNotebook  ', e as any)
+      throw e
+    }
+    const notebook = res.notebook
+
+    if (!notebook) {
+      return this.printCell('⚠️ __Error__: no cells found!')
+    }
+    return notebook as any as Serializer.Notebook // ugly cast :(
```

Comment:
Are you missing logic ala https://github.com/stateful/vscode-runme/blob/127028fb126a1512e0ac4dec84df0ba83db5f566/src/extension/serializer.ts#L296 
To guess the language id of the cells?

Reply:
Now that we are just going with a single class and just changing the library can we eliminate most of this diff?